src/native/libs/System.Security.Cryptography.Native.Apple/pal_swiftbindings.swift

```diff
@@ -350,3 +359,285 @@ public func AppleCryptoNative_HKDFDeriveKey(
+            return Unmanaged.passRetained(box).toOpaque()
+        default:
+            pcbDigest.pointee = 0
+            return nil
+    }
+}
+
+@_silgen_name("AppleCryptoNative_DigestUpdate")
+@available(iOS 13, tvOS 13, macOS 13, *)
+public func AppleCryptoNative_DigestUpdate(ctx: UnsafeMutableRawPointer?, pBuf: UnsafeMutableRawPointer?, cBuf: Int32) -> Int32 {
+    if cBuf == 0 {
+        return 1
+    }
+
+    guard let ctx, let pBuf, cBuf >= 0 else {
+        return -1
+    }
+
+    let box = Unmanaged<HashBox>.fromOpaque(ctx).takeUnretainedValue()
+    let source = Data(bytesNoCopy: pBuf, count: Int(cBuf), deallocator: Data.Deallocator.none)
+
+    switch box.algorithm {
+        case .md5:
+            var hash = box.value as! Insecure.MD5
+            hash.update(data: source)
+            box.value = hash
+            return 1
+        case .sha1:
+            var hash = box.value as! Insecure.SHA1
+            hash.update(data: source)
+            box.value = hash
+            return 1
+        case .sha256:
+            var hash = box.value as! SHA256
+            hash.update(data: source)
+            box.value = hash
+            return 1
+        case .sha384:
+            var hash = box.value as! SHA384
+            hash.update(data: source)
+            box.value = hash
+            return 1
+        case .sha512:
+            var hash = box.value as! SHA512
+            hash.update(data: source)
+            box.value = hash
+            return 1
+        default:
+            return -1
+    }
+}
+
+@_silgen_name("AppleCryptoNative_DigestReset")
+@available(iOS 13, tvOS 13, macOS 13, *)
+public func AppleCryptoNative_DigestReset(ctx: UnsafeMutableRawPointer?) -> Int32 {
+    guard let ctx else {
+        return -1
+    }
+
+    let box = Unmanaged<HashBox>.fromOpaque(ctx).takeUnretainedValue()
+
+    switch box.algorithm {
+        case .md5:
+            box.value = Insecure.MD5()
+            return 1
+        case .sha1:
+            box.value = Insecure.SHA1()
+            return 1
+        case .sha256:
+            box.value = SHA256()
+            return 1
+        case .sha384:
+            box.value = SHA384()
+            return 1
+        case .sha512:
+            box.value = SHA512()
+            return 1
+        default:
+            return -2
+    }
+}
+
+@_silgen_name("AppleCryptoNative_DigestFinal")
+@available(iOS 13, tvOS 13, macOS 13, *)
+public func AppleCryptoNative_DigestFinal(ctx: UnsafeMutableRawPointer?, pOutput: UnsafeMutablePointer<UInt8>?, cbOutput: Int32) -> Int32 {
+    guard let ctx, let pOutput else {
+        return -1
+    }
+
+    let box = Unmanaged<HashBox>.fromOpaque(ctx).takeUnretainedValue()
+    let destination = UnsafeMutableRawBufferPointer(start: pOutput, count: Int(cbOutput))
+
+    switch box.algorithm {
+        case .md5:
+            let hash = (box.value as! Insecure.MD5).finalize()
+            _ = hash.withUnsafeBytes { digest in digest.copyBytes(to: destination) }
```

Comment:
I believe we'll have proper size guards in the C# layer, but what would happen here if we got that wrong and `destination` was undersized?  (In C, the answer would be an OOB write, so I'm fine with any answer here... I'm just trying to learn Swift conventions one piece at a time)

Reply:
This was easy to check so added checks.